### ISSUE TITLE
Adjoint solves coming from problems with FROSch/MueLu preconditioners on Blake

### ISSUE THREAD

**Issue by thartland** [sandialabs/Albany#947]
I've been having trouble with PyAlbany on Blake, specifically applying the Hessian to vectors wherein the Newton linear solves of the forward model use either FROSch or MueLu preconditioners.

I've followed the gcc build instructions provided under Albany/doc/dashboards/blake.sandia.gov, specifically using modules specified in blake_gcc_modules.sh. I then built Trilinos on the develop branch (pulled on the morning of May 7th 2023). Using the build script do-cmake-trilinos-gcc-serial.

I then build Albany (also pulled on morning of May 7th, 2023) using do-cmake-albany-gcc-serial, with the only modification being that the TRILINOS install dir points my Trilinos install.

What I find is that Albany builds fine and furthermore the PyAlbany tests all pass. I then try to run a PyAlbany code that requires the action of the Hessian, wherein I use either MueLu or FROSch for the Newton linear system preconditioner, see e.g, [input_fo_gis_unstructT_FROSch.yaml](https://github.com/sandialabs/Albany/blob/master/tests/landIce/FO_GIS/input_fo_gis_unstructT_FROSch.yaml#L197). 

When I run a PyAlbany code which calls an input file wherein explicit matrix transpose is enabled
```
  Piro:
    Sensitivity Method: Adjoint
    Enable Explicit Matrix Transpose: true
```
the code still fails on the adjoint solves with error indicating that transpose has not been enabled.

```
MueLu::XpetraOperator does not support applying the adjoint operator
Traceback (most recent call last):
  File "HessianCompression.py", line 69, in <module>
    main(parallelEnv)
  File "HessianCompression.py", line 46, in main
    eigVals, eigVecs = singlePass(Hess, r)
  File "/ascldap/users/thartla/albany/Albany2023_04_26/Albany/builds/Blake/pyAlbany/PyAlbany/RandomizedCompression.py", line 58, in singlePass
    y = Op.dot(omega)
  File "HessianCompression.py", line 18, in dot
    me.problem.performSolve()
RuntimeError: /ascldap/users/thartla/trilinos/Trilinos2023_04_26/Trilinos/packages/muelu/src/Operators/MueLu_XpetraOperator_def.hpp:83:

Throw number = 1

Throw test that evaluated to true: mode!=Teuchos::NO_TRANS
```

It should be noted that all PyAlbany tests use Ifpack2 preconditioners e.g., [the randomized compression PyAlbany tests](https://github.com/sandialabs/Albany/blob/master/pyAlbany/tests/randCompress/input_conductivity_dist_paramT.yaml#L160) and so this is why the PyAlbany tests pass, while I am still getting errors as explained above.

**Comment by ikalash** [comment]
Since this concerns PyAlbany, @kliegeois may be the best one to try triage/resolve this.

**Comment by kliegeois** [comment]
Thanks @thartland for reporting this and sorry that it happened. Can you provide a small PyAlbany code snippet that uses the yaml file listed above and that reproduces the error?

**Comment by thartland** [comment]
No worries @kliegeois and thanks for the quick response. It might be easier to use the [input_fo_humboldt_frosch_fluxdiv.yaml](https://github.com/sandialabs/Albany/blob/master/tests/landIce/FO_GIS/input_fo_humboldt_frosch_fluxdiv.yaml), as it doesn't require the mesh to be populated. The following will reproduce the above described error.

```
import numpy as np
from mpi4py import MPI
from PyAlbany import Utils
from PyAlbany import AlbanyInterface as pa
from PyAlbany.RandomizedCompression import singlePass


class Hessian:
   def __init__(me, problem, parameterIndex, responseIndex):
       me.problem = problem
       me.parameterIndex = parameterIndex
       me.responseIndex  = responseIndex
       me.Map            = me.problem.getParameterMap(me.parameterIndex)
   def dot(me, x):
       me.problem.setDirections(me.parameterIndex, x)
       me.problem.performSolve()
       return me.problem.getReducedHessian(me.responseIndex, me.parameterIndex)


def main(parallelEnv):
    comm = parallelEnv.getComm()
    myGlobalRank = comm.getRank()
    iAmRoot = myGlobalRank == 0

    # Create an Albany problem:
    fileName = 'input_fo_humboldt_frosch_fluxdiv.yaml'
    pList    = Utils.createParameterList(fileName, parallelEnv)
    problem = Utils.createAlbanyProblem(pList, parallelEnv)
    problem.performSolve()

    parameterIndex = 0
    responseIndex  = 0
    parameterMap = problem.getParameterMap(parameterIndex)
 
    Hess = Hessian(problem, parameterIndex, responseIndex)
    k = 1
    p = 0
    r = k + p
    eigVals, eigVecs = singlePass(Hess, r)



if __name__ == '__main__':
    parallelEnv = Utils.createDefaultParallelEnv()
    main(parallelEnv)

                              
```

**Comment by kliegeois** [comment]
Thanks @thartland ! I will look into this tomorrow and potentially create a unit test to prevent this from happening again.

**Comment by mperego** [comment]
Thanks @thartland, and sorry for not being able to help while on travel.
 @kliegeois  I think it's because PyAlbany does not pass the adjoint model when creaing the solver.
https://github.com/sandialabs/Albany/blob/master/pyAlbany/src/Albany_Interface.cpp#L116-L117

It should be something like
https://github.com/sandialabs/Albany/blob/master/src/Main_Analysis.cpp#L94-L105

**Comment by kliegeois** [comment]
Thanks @mperego ! That's my guess too ( I am facing the exact same issue when I am working with the transient branch with Piro).

**Comment by kliegeois** [comment]
@thartland the issue has been closed automatically with the merge of PR #948 .
Could you verify that you don't face the issue anymore?
If needed, I can reopen the issue.
Thanks!
